config_api.go

```diff
@@ -17,9 +20,47 @@ type ClientBinary struct {
 }
 
 type ClientConfig struct {
-	ClientBinaries   []ClientBinary `yaml:"client_binaries" json:"client_binaries"`
-	S3ClientBinaries []ClientBinary `yaml:"s3_client_binaries" json:"s3_client_binaries"`
-	LatestRevision   string         `yaml:"latest_revision" json:"latest_revision"`
+	ClientBinaries   []ClientBinary
+	S3ClientBinaries []ClientBinary
+	LatestRevision   string
+	S3URLPrefix      string
+}
+
+func (c *ClientConfig) populateClientBinaries(ctx context.Context, bucket pail.Bucket, prefix string) error {
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+
+	iter, err := bucket.List(ctx, prefix)
+	if err != nil {
+		return errors.Wrap(err, "listing client bucket")
+	}
+	for iter.Next(ctx) {
+		item := strings.TrimPrefix(iter.Item().Name(), prefix)
+		name := strings.Split(item, "/")
+		if len(name) != 2 || !strings.Contains(name[1], "evergreen") {
+			continue
+		}
+		displayName, ok := ValidArchDisplayNames[name[0]]
+		if !ok {
+			continue
+		}
+		osArchParts := strings.Split(name[0], "_")
+		if len(osArchParts) != 2 {
```

Comment:
nit: why do we skip if we find an underscore / can we make a quick comment explaining?

Reply:
The truth is, that check is kind of redundant given the previous check that the name is a key in the `ValidArchDisplayNames` map. I've removed it.
The idea of the check has been to make sure the name conformed to the convention that it have two parts (os and arch) separated by an underscore.